### DISCUSSION TITLE
Dependency Injection Pattern (state init with props) breaks Typesafety

### DISCUSSION THREAD

**Discussion by Twiggeh** [pmndrs/zustand#1629] [General] [answered]
## Summary
When following the pattern on [zustand documentation](https://docs.pmnd.rs/zustand/guides/initialize-state-with-props) typesafety can be broken.

E.g. when you have a store with type 
```ts
{ bears: number }
```
And you pass to `initProps` 
```ts
{ bears: undefined }
```
`bears` will become `undefined`, but the type says its `number` !  

## Link to reproduction
[Link](https://codesandbox.io/s/serverless-wave-22zqbj?file=/src/store.ts)

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [x] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment, marked as answer]
You need to enable `exactOptionalPropertyTypes` in tsconfig, if you want catch it. I don't think we can do anything from the library perspective.

https://tsplay.dev/wRBG1w

<img width="551" alt="image" src="https://user-images.githubusercontent.com/490574/218450834-348f5ed8-9d05-45b6-ae37-2137570945f4.png">

**Reply by Twiggeh** [reply]
I had created something like this to solve it (in this case the default props signify that those properties should always be defined in store, but are optional in a prop):

### Store
```ts
import { create } from "zustand";

type Props = {
	fishes?: string;
	bears: number;
	honey?: boolean;
};

export const DEFAULT_PROPS = {
	fishes: "",
	bears: 12,
} satisfies Props;

type Store = { addBear: (amount?: number) => void } & Props &
	typeof DEFAULT_PROPS;

export const injectStore = (initProps?: Props & typeof DEFAULT_PROPS) => {
	return create<Store>()((set, get) => {
		return {
			...DEFAULT_PROPS,
			...initProps,
			addBear: (amount = 1) => set({ bears: get().bears + amount }),
		};
	});
};
```

### App
```ts
import { injectStore, DEFAULT_PROPS } from "./store";

function App() {
  const injectedStore = useRef(
    injectStore({ ...DEFAULT_PROPS, honey: false, bears: 12 })
    ).current;
```

But your solution works well too !

Only for enabling that option in TS you would need to go back to any optional property where you also expected `undefined` and add the `| undefined`, whereas with mine you would need to export the defaults and always spread them. 

Maybe we can at least add a "common pitfall" section and put the compiler option there ?

**Reply by dai-shi** [reply]
This is just how TS works and not directly related with zustand. We would probably like to have a best practice or tutorial docs for zustand usage with typescript. There, it might be possible to have a note about the option. I think we are still under the discussion about how such docs should be.

**Reply by Twiggeh** [reply]
> We would probably like to have a best practice or tutorial docs for zustand usage with typescript.

Yes please do!

I got bamboozled by some properties suddenly turning into undefined :D

**Reply by dai-shi** [reply]
cc: docs team @sewera @chrisk-7777 @dbritto-dev

**Reply by chrisk-7777** [reply]
Not a problem, addressed here: https://github.com/pmndrs/zustand/pull/1634

@Twiggeh Do you think this would have been enough if you were approaching this problem fresh? In particular these added lines: https://github.com/pmndrs/zustand/pull/1634/files#diff-ebbd2bc944d3787f96778ca0215221865d6ad488c3ad305cd62d39cee7699d27R499

**Reply by Twiggeh** [reply]
Yes thank you :heart:
